toolchain/semantics/semantics_ir_factory.cpp

```diff
@@ -165,91 +65,210 @@ static auto GetBinaryOp(TokenKind kind) -> Semantics::BinaryOperator::Op {
+  for (auto it = range.begin();; ++it) {
+    auto parse_node = *it;
+    int32_t subtree_size = parse_tree().node_subtree_size(parse_node);
+    switch (auto parse_kind = parse_tree().node_kind(parse_node)) {
+      case ParseNodeKind::CodeBlock(): {
+        // Just merges children.
+        llvm::SmallVector<Semantics::NodeRef, 0> sem_ir;
+        while (subtree_size > 1) {
+          subtree_size -=
+              parse_tree().node_subtree_size(node_stack.back().parse_node);
+          sem_ir.insert(sem_ir.begin(), node_stack.back().sem_ir.begin(),
+                        node_stack.back().sem_ir.end());
+          node_stack.pop_back();
+        }
+        node_stack.push_back(
+            TraversalStackEntry(parse_node, std::move(sem_ir)));
+        break;
+      }
+      case ParseNodeKind::DeclaredName(): {
+        RequireNodeEmpty(parse_node, parse_kind, subtree_size);
+        // DeclaredNames will be acted upon as part of the declaring construct.
+        node_stack.push_back(TraversalStackEntry(parse_node));
+        break;
+      }
+      case ParseNodeKind::FunctionDeclaration(): {
+        // Currently we only have definitions, so this is a CodeBlock.
+        subtree_size -=
+            parse_tree().node_subtree_size(node_stack.back().parse_node);
+        llvm::SmallVector<Semantics::NodeRef, 0> body =
+            std::move(node_stack.back().sem_ir);
+        node_stack.pop_back();
+
+        // Next is the FunctionSignature.
+        subtree_size -=
+            parse_tree().node_subtree_size(node_stack.back().parse_node);
+        CARBON_CHECK(subtree_size == 1);
+        llvm::SmallVector<Semantics::NodeRef, 0> sig =
+            std::move(node_stack.back().sem_ir);
+        node_stack.pop_back();
+
+        // TODO: This replacement is in particular why I want to change
+        // the IR setup now, but for now I want to just produce output that
+        // satisfies tests without changes.
+        auto orig_function = semantics_.nodes_.Get<Semantics::Function>(sig[0]);
+        auto orig_set_name = semantics_.nodes_.Get<Semantics::SetName>(sig[1]);
+        llvm::SmallVector<Semantics::NodeRef, 0> sem_ir;
+        auto function_id = next_id();
+        sem_ir.push_back(semantics_.nodes_.Store(
+            Semantics::Function(orig_function.node(), function_id, body)));
+        sem_ir.push_back(semantics_.nodes_.Store(Semantics::SetName(
+            orig_set_name.node(), orig_set_name.name(), function_id)));
+
+        node_stack.push_back(
+            TraversalStackEntry(parse_node, std::move(sem_ir)));
+        break;
+      }
+      case ParseNodeKind::FunctionSignature(): {
+        // TODO: Skip over the parameter list for now.
+        subtree_size -=
+            parse_tree().node_subtree_size(node_stack.back().parse_node);
+        node_stack.pop_back();
+
+        // TODO: At this point, it should be possible to forward-declare the
+        // function so that it can be called a code block. For now, we just
```

Comment:
```suggestion
        // function so that it can be called from its code block. For now, we just
```
Is this what you mean?